testing/test_collection.py

```diff
@@ -1404,3 +1404,17 @@ def a(): return 4
     result = testdir.runpytest()
     # Not INTERNAL_ERROR
     assert result.ret == ExitCode.INTERRUPTED
+
+
+def test_does_not_crash_on_recursive_symlink(testdir: Testdir) -> None:
+    """Regression test for an issue around recursive symlinks (#7951)."""
+    testdir.makepyfile(
+        """
+        def test_foo(): assert True
+        """
+    )
+    os.symlink("recursive", str(testdir.tmpdir.join("recursive")))
```

Comment:
Another comment: should use `symlink_or_skip` here (already imported), for platforms which don't support symlinks.